app/views/service_worker/index.js.erb

```diff
@@ -1,144 +1,132 @@
+    headers: {'Content-Type': 'text/html; charset=utf-8'}
+  });
+}
 
-function onPush(event) {
-  var title = 'DEV Connect 👋 ';
-  var body = event.data.text();
-  var icon = '<%= image_path "devlogo-pwa-128.png" %>';
-  var tag = 'default-tag' + body;
-  event.waitUntil(
-    self.registration.showNotification(title, {
-      body: body,
-      icon: icon,
-      tag: tag,
-      data: {
-        url: 'https://dev.to/connect'
+self.addEventListener('fetch', event => {
+  const url = new URL(event.request.url);
+  if (url.origin === location.origin) {
+    // 
+    if (event.clientId === "" && // Not fetched via AJAX after page load.
+      event.request.method == "GET" && // Don't fetch on POST, DELETE, etc.
+      !event.request.referrer.includes('/signout_confirm') && // If this is the referrer, we instead want to flush.
+      !url.href.includes('i=i') && // Parameter representing "internal" navigation.
+      !url.href.includes('?signin') && // Don't run on sign in.
+      !url.href.includes('/onboarding') && // Don't run on onboarding.
+      !url.href.includes('/api/') && // Don't run on API endpoints.
+      !url.href.includes('/users/auth') && // Don't run on authentication.
+      !url.href.includes('/shell_') && // Don't fetch for shell.
+      !url.href.includes('/admin') && // Don't fetch for administrate dashboard.
+      !url.href.includes('/internal') && // Don't fetch for internal dashboard.
+      caches.match('/shell_top') && // Ensure shell_top is in the cache.
+      caches.match('/shell_bottom')) { // Ensure shell_bottom is in the cache.
+      event.respondWith(createPageStream(event.request)); // Respond with the stream
+
+      // Ping version endpoint to see if we should fetch new shell. 
+      if (!caches.match('/async_info/shell_version')) { // Check if we have a cached shell version
+        caches.open(staticCacheName)
+        .then(cache => cache.addAll([
+          "/async_info/shell_version",
+        ]))  
+        return
       }
-    })
-  );
-};
+      fetch('/async_info/shell_version').then(response => response.json()).then(json => {
+        caches.match('/async_info/shell_version').then(cachedResponse => cachedResponse.json()).then(cacheJson => {
+          if (cacheJson['version'] != json['version']) {
+            caches.open(staticCacheName)
+            .then(cache => cache.addAll([
+              "/shell_top",
+              "/shell_bottom",
+              "/async_info/shell_version"
+            ]))
+          }
+        })
+      });
+      return;
+    }
 
-function onNotificationClick(event) {
-  if (Notification.prototype.hasOwnProperty('data')) {
-    var url = event.notification.data.url;
-    event.waitUntil(clients.openWindow(url));
+    // Fetch new shell upon events that signify change in session.
+    if (event.clientId === "" &&
+      (event.request.referrer.includes('/signout_confirm') || url.href.includes('?signin') || url.href.includes('/onboarding') || Math.random() > 0.9)) {
```

Comment:
why the random?

Reply:
I just thought it might be responsible to occasionally fetch a new shell regardless of circumstances _just in case_. I'm not married to that idea though.